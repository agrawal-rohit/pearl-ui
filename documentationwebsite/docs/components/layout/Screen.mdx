---
sidebar_position: 3
title: Screen
---

Screen is a layout component that you can use to wrap all the views in your app. By default, it renders a [Box](./box) element.

The Screen component also implements a `SafeAreaView` under the hood. This allows the content wrapped by the Screen component to
always be within the safe areas of the device. (This is especially helpful for keeping your app accessible on devices with rounded corners
or camera notches)

## Import

```jsx
import { Screen } from "pearl-ui";
```

## Usage

```jsx
<Screen>This is the screen</Screen>
```

## Example

import Snack from "../../../src/components/ExpoSnack";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

<Tabs
  defaultValue="light"
  values={[
    { label: 'Light', value: 'light', },
    { label: 'Dark', value: 'dark', },
  ]
}>
<TabItem value="light">

<Snack snackId="@agrawal-rohit/screen-light" />

</TabItem>

<TabItem value="dark">

<Snack snackId="@agrawal-rohit/screen-dark" />

</TabItem>
</Tabs>

## Default Component Style

```json
{
  "baseStyle": {
    "backgroundColor": {
      "light": "neutral.100",
      "dark": "neutral.800"
    },
    "padding": "m"
  },
  "sizes": {},
  "variants": {},
  "defaults": {}
}
```
