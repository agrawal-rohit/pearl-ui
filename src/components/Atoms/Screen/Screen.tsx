import React from "react";
import Box from "../Box/Box";
import { SafeAreaView, StatusBar, Platform } from "react-native";
import { BoxProps, useRestyle, boxRestyleFunctions } from "@shopify/restyle";
import { Theme } from "../../../theme/theme";

type ScreenProps = BoxProps<Theme>;

const Screen: React.FC<ScreenProps> = ({ children, ...rest }) => {
  const props = useRestyle(boxRestyleFunctions, rest);

  return (
    <Box
      padding="m"
      backgroundColor="mainBackground"
      flex={1}
      flexDirection="column"
      {...props}
    >
      <SafeAreaView
        style={{
          flex: 1,
          paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,
        }}
      >
        {children}
      </SafeAreaView>
    </Box>
  );
};

export default Screen;
