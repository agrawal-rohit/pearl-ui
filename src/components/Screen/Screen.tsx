import React from "react";
import { Box } from "../..";
import { SafeAreaView, StatusBar, Platform } from "react-native";

interface ScreenProps {}

const Screen: React.FC<ScreenProps> = ({ children }) => {
  return (
    <Box
      padding="m"
      backgroundColor="mainBackground"
      flex={1}
      flexDirection="column"
    >
      <SafeAreaView
        style={{
          flex: 1,
          paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,
        }}
      >
        {children}
      </SafeAreaView>
    </Box>
  );
};

export default Screen;
