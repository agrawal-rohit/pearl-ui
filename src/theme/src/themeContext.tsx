import AppLoading from "expo-app-loading";
import {
  useFonts,
  Poppins_300Light,
  Poppins_400Regular,
  Poppins_500Medium,
  Poppins_600SemiBold,
  Poppins_700Bold,
  Poppins_800ExtraBold,
} from "@expo-google-fonts/poppins";
import React, { useEffect, useState, createContext } from "react";
import { useColorScheme } from "react-native";
import { baseTheme } from "./basetheme";
import { BasePearlTheme } from "./types";

type ThemeType = "light" | "dark" | "system";

export interface IThemeContext {
  /** Theme configuration object for the active color mode */
  theme: BasePearlTheme;
  /** Active color mode */
  colorMode: ThemeType;
  /** Function to toggle the active color mode */
  toggleColorMode(): void;
}

interface ThemeProviderProps {
  /** Initial color mode for the theme (light, dark, system) */
  initialColorMode?: ThemeType;
  /** The theme configuration object */
  theme?: BasePearlTheme;
  /** A mapping object of the fonts used by the theme */
  fonts?: object;
  /**React children */
  children: React.ReactNode;
}

const baseFonts = {
  "Poppins-Light": Poppins_300Light,
  "Poppins-Regular": Poppins_400Regular,
  "Poppins-Medium": Poppins_500Medium,
  "Poppins-SemiBold": Poppins_600SemiBold,
  "Poppins-Bold": Poppins_700Bold,
  "Poppins-ExtraBold": Poppins_800ExtraBold,
};

export const themeContext = createContext({} as IThemeContext);

/**
 * The main provider component which controls the theme and color mode of the components in the entire application.
 */
export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  initialColorMode = "light",
  theme = baseTheme,
  fonts = baseFonts,
  children,
}) => {
  const [colorMode, setColorMode] = useState<ThemeType>(initialColorMode);
  const systemThemeStyle = useColorScheme() as ThemeType;
  const [haveFontsLoaded] = useFonts(fonts);

  const changeThemeTo = (colorMode: ThemeType) => {
    if (colorMode === "dark") {
      setColorMode("dark");
    } else {
      setColorMode("light");
    }
  };

  const toggleColorMode = () => {
    if (colorMode === "light") {
      changeThemeTo("dark");
    } else {
      changeThemeTo("light");
    }
  };

  useEffect(() => {
    // Set initial Theme
    if (colorMode === "light") {
      changeThemeTo("light");
    } else if (colorMode === "dark") {
      changeThemeTo("dark");
    } else if (colorMode === "system") {
      changeThemeTo(systemThemeStyle);
    }
  }, [colorMode, initialColorMode, theme]);

  if (haveFontsLoaded) {
    return (
      <themeContext.Provider
        value={{ theme, colorMode, toggleColorMode } as IThemeContext}
      >
        {children}
      </themeContext.Provider>
    );
  }

  return <AppLoading />;
};
