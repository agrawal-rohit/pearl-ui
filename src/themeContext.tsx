import React, { useEffect, useState, useContext, createContext } from "react";
import { useColorScheme } from "react-native";
import { ThemeProvider as RestyleThemeProvider } from "@shopify/restyle";
import { baseTheme, darkTheme, Theme } from "./theme";
import { ReactElement } from "react-color/node_modules/@types/react";

type ThemeType = "light" | "dark" | "system";

interface IThemeContext {
  themeMode: ThemeType;
  toggleTheme(theme: string): void;
}

interface ThemeProviderProps {
  defaultMode?: ThemeType;
  light?: Theme;
  dark?: Theme;
}

export const themeContext = createContext({} as IThemeContext);

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  defaultMode = "light",
  light = baseTheme,
  dark = darkTheme,
  children,
}) => {
  const [themeMode, setThemeMode] = useState<ThemeType>(defaultMode);
  const systemThemeStyle = useColorScheme() as ThemeType;

  const toggleTheme = () => {
    if (themeMode === "light") {
      setThemeMode("dark");
    } else {
      setThemeMode("light");
    }
  };

  useEffect(() => {
    // Set theme
    if (defaultMode === "system") {
      setThemeMode(systemThemeStyle);
    }
  }, []);

  return (
    <themeContext.Provider value={{ themeMode, toggleTheme } as IThemeContext}>
      <RestyleThemeProvider theme={themeMode === "light" ? light : dark}>
        {children}
      </RestyleThemeProvider>
    </themeContext.Provider>
  );
};

export const useTheme = () => useContext(themeContext);
