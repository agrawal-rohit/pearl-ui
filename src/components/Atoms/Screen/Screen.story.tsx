import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "./Screen";

storiesOf("Screen", module)
  .addDecorator((getStory) => getStory())
  .add("Screen", () => <Screen></Screen>);
