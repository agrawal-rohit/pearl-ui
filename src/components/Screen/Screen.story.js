import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";

storiesOf("Screen", module)
  .addDecorator((getStory) => getStory())
  .add("Screen", () => <Screen></Screen>);
