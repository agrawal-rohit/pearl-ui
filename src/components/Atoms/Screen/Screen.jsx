import React from "react";
import Box from "../Box/Box";
import { SafeAreaView, StatusBar, Platform } from "react-native";
import { useRestyle, boxRestyleFunctions } from "@shopify/restyle";
import { IBasePearlTheme } from "../../../theme/src/types";

// type ScreenProps = BoxProps<IBasePearlTheme>;

const Screen = ({ children, ...rest }) => {
  const props = useRestyle(boxRestyleFunctions, rest);

  return (
    <Box
      padding="m"
      backgroundColor="mainBackground"
      flex={1}
      flexDirection="column"
      {...props}
    >
      <SafeAreaView
        style={{
          flex: 1,
          paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,
        }}
      >
        {children}
      </SafeAreaView>
    </Box>
  );
};

export default Screen;
