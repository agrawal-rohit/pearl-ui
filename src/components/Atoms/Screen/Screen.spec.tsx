import React from "react";
import Screen from "./Screen";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "../../../themeContext";

jest.useFakeTimers();

describe("Atoms/Screen", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <Screen>asdasd</Screen>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
