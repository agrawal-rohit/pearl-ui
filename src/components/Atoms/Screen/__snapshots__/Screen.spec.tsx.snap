// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Screen passes the snapshot test 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingTop": 0,
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    bounces={true}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={true}
    enableResetScrollToCoords={true}
    extraHeight={100}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    scrollable={true}
    showScrollBar={false}
    showsVerticalScrollIndicator={false}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "display": "flex",
            "flex": 1,
            "flexDirection": "column",
            "padding": 16,
          }
        }
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
